File: apps/api/go_reading_oasis/services/books_service.go
package services

import (
	"fmt"

	"github.com/bwe/bwe/apps/api/go_reading_oasis/data"
	"github.com/bwe/bwe/apps/api/go_reading_oasis/models"
)

// Gets a list of books from the database.
func ListBooks(pagination models.Pagination) (*models.ListBooksResponse, models.Error) {
	books, err := data.GetBooks(&pagination)
	if err != nil {
		return nil, err.Error()
	}
	count := len(books)
	return &models.ListBooksResponse{
		Data:  books,
		Page:  &pagination.Page,
		Limit: &pagination.Limit,
		Count: &count,
	}, nil
}

// Creates a book.
func CreateBook(request models.CreateBookRequest) (*models.CreateBookResponse, models.Error) {
	book := models.Book{Author: request.Author, Title: request.Title}
	if _, err := data.CreateBook(&book); err != nil {
		return nil, err.Error()
	}
	return &models.CreateBookResponse{Data: book}, nil
}

// Gets a book from the database by its ID.
//
// Returns the GetBookResponse with the book as the "Data", whether the book was not found, and any
// errors that occurred while performing the operation.
func GetBook(id int) (*models.GetBookResponse, models.Error) {
	book, err := data.GetBook(id)
	if err != nil {
		return nil, getBookNotFoundError(id)
	}
	return &models.GetBookResponse{Data: *book}, nil
}

// Updates the book with the given ID in the database.
//
// Returns the update book response with the book value as data, whether the book was not found
// (true if it was not found) in the database, and any errors that occured while performing the
// operation.
func UpdateBook(id int, request models.UpdateBookRequest) (*models.UpdateBookResponse, bool, models.Error) {
	book, err := data.GetBook(id)
	if err != nil {
		return nil, true, getBookNotFoundError(id)
	}
	book.Author = request.Author
	book.Title = request.Title
	if err := data.UpdateBook(book); err != nil {
		return nil, false, err.Error()
	}
	return &models.UpdateBookResponse{Data: *book}, false, nil
}

// Deletes the book with the given id in the database.
//
// Returns whether the book was not found and any errors that occured during the operation.
func DeleteBook(id int) models.Error {
	book, err := data.GetBook(id)
	if err != nil {
		return getBookNotFoundError(id)
	}
	if err := data.DeleteBook(book); err != nil {
		return err.Error()
	}
	return nil
}

func getBookNotFoundError(id int) models.Error {
	return models.ApiError{Type: models.NotFoundError,
		Message: fmt.Sprintf("Book with ID [%v] was not found.", id)}
}

File: apps/api/py_reading_oasis/services/books_service.py
from apps.api.py_reading_oasis.database.database import Book
from apps.api.py_reading_oasis.exceptions.bad_request_error import BadRequestError
from apps.api.py_reading_oasis.exceptions.resource_not_found_error import ResourceNotFoundError
from apps.api.py_reading_oasis.models.direction import Direction
from apps.api.py_reading_oasis.repository import books_repository


def get_book(book_id):
    return get_book_by_id(book_id).serialize()


def get_books(pagination):
    if not _validate_book_sort(pagination.order, pagination.sort_field):
        raise BadRequestError("'{}' is not a valid book field.".format(pagination.sort_field))
    order = Book.created_at.desc()
    if pagination.order == Direction.ascending:
        order = Book.created_at.asc()
    books = books_repository.get_books()
    return [book.serialize() for book in books]


def create_book(request):
    book = Book(author=request.get('author'), title=request.get('title'))
    return books_repository.create_book(book).serialize()


def update_book(book_id, request):
    book = get_book_by_id(book_id)
    author = request.get('author')
    title = request.get('title')
    return books_repository.update_book(book, author, title).serialize()


def delete_book(book_id):
    book = get_book_by_id(book_id)
    books_repository.delete_book(book)


def get_book_by_id(book_id):
    book = books_repository.get_book(book_id)
    if book is None:
        raise ResourceNotFoundError("No book found with id '{}'".format(book_id))
    return book


def _validate_book_sort(direction, sort_field):
    if direction == Direction.ascending or direction == direction.descending and str(
            sort_field).strip() in Book.fields():
        return True
    return False
